### DISCUSSION TITLE
ScriptReference/UIElements.VisualElement-generateVisualContent

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#86] [Notes]
# ScriptReference/UIElements.VisualElement-generateVisualContent

https://docs.unity3d.com/ScriptReference/UIElements.VisualElement-generateVisualContent.html

**Comment by vertxxyz** [comment]
If nothing is being drawn, check that:

- Vertices are drawn at the correct depth. The helper [`Vertex.nearZ`](https://docs.unity3d.com/ScriptReference/UIElements.Vertex-nearZ.html) provides this z depth for you.
- The correct triangle winding order is used. This is clockwise[^1].
- Colours used for vertices are not transparent.
- Your triangles have area. If you are drawing lines, two vertices isn't enough, you will have to manually give them volume as triangles.
- Your element isn't hidden behind others.

If these are not your problem, start with the simple case (drawing a **large** triangle/quad) in a basic element and work your way upwards to drawing complex content.

[^1]: The coordinate space starts in the top left of the element, with positive coordinates trending to the bottom right

**Comment by vertxxyz** [comment]
As of 2022.1 you should generally use [`painter2D`](https://docs.unity3d.com/2023.2/Documentation/ScriptReference/UIElements.MeshGenerationContext-painter2D.html) to draw content. You do not need to call `Clear` on it before usage.
It has a much more straight forward API, is harder to abuse, and has fairly good error reporting.
```csharp
/// <summary>
/// Draws a 1px wide circle aligned in the center of the element.
/// </summary>
private static void GenerateVisualContent(MeshGenerationContext obj)
{
	const int pixelWidth = 1;
	Rect dimensions = obj.visualElement.contentRect;
	float radius = Math.Min(dimensions.width, dimensions.height) * 0.5f - pixelWidth ;

	Painter2D p2d = obj.painter2D;
	p2d.lineWidth = pixelWidth;
	p2d.strokeColor = Color.white;
	p2d.BeginPath();
	p2d.Arc(dimensions.center, radius, 0, Angle.Turns(1));
	p2d.Stroke();
}
```
